IdentityCore/src/cache/ios/MSIDKeychainTokenCache.m

```diff
@@ -54,7 +242,181 @@ - (BOOL)removeItemWithKey:(MSIDTokenCacheKey *)key
                                context:(id<MSIDRequestContext>)context
                                  error:(NSError **)error
 {
-    return nil;
+    assert(serializer);
+    
+    MSID_LOG_INFO(context, @"Get keychain items, key info (account: %@ service: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service));
+    MSID_LOG_INFO_PII(context, @"Get keychain items, key info (account: %@ service: %@)", key.account, key.service);
+    
+    NSMutableDictionary *query = [self defaultQuery];
+    if (key.service)
+    {
+        [query setObject:key.service forKey:(id)kSecAttrService];
+    }
+    if (key.account)
+    {
+        [query setObject:key.account forKey:(id)kSecAttrAccount];
+    }
+    [query setObject:@YES forKey:(id)kSecReturnData];
+    [query setObject:@YES forKey:(id)kSecReturnAttributes];
+    [query setObject:(id)kSecMatchLimitAll forKey:(id)kSecMatchLimit];
+    
+    CFTypeRef cfItems = nil;
+    MSID_LOG_INFO(context, @"Trying to find keychain items...");
+    OSStatus status = SecItemCopyMatching((CFDictionaryRef)query, &cfItems);
+    MSID_LOG_INFO(context, @"Keychain find status: %d", status);
+    
+    if (status == errSecItemNotFound)
+    {
+        return @[];
+    }
+    else if (status != errSecSuccess)
+    {
+        if (error)
+        {
+            *error = [[NSError alloc] initWithDomain:NSOSStatusErrorDomain code:status userInfo:@{NSLocalizedDescriptionKey:@"Failed to get items from keychain."}];
+        }
+        MSID_LOG_ERROR(context, @"Failed to find keychain item (status: %d)", status);
+        return nil;
+    }
+    
+    NSArray *items = CFBridgingRelease(cfItems);
+    
+    NSMutableArray *tokenItems = [[NSMutableArray<MSIDToken *> alloc] initWithCapacity:items.count];
+    for (NSDictionary *attrs in items)
+    {
+        NSData *itemData = [attrs objectForKey:(id)kSecValueData];
+        MSIDToken *tokenItem = [serializer deserialize:itemData];
+        
+        if (tokenItem)
+        {
+            // Delete tombstones generated from previous versions of ADAL.
+            if (tokenItem.tokenType == MSIDTokenTypeRefreshToken && [tokenItem.token isEqualToString:@"<tombstone>"])
+            {
+                NSString *service = attrs[(id)kSecAttrService];
+                NSString *account = attrs[(id)kSecAttrAccount];
+                
+                if (!service || !account)
+                {
+                    continue;
+                }
+             
+                NSMutableDictionary *deleteQuery = [self defaultQuery];
+                [deleteQuery setObject:service forKey:(id)kSecAttrService];
+                [deleteQuery setObject:account forKey:(id)kSecAttrAccount];
+                
+                MSID_LOG_INFO(context, @"Trying to delete tombstone item...");
+                OSStatus status = SecItemDelete((CFDictionaryRef)deleteQuery);
+                MSID_LOG_INFO(context, @"Keychain delete status: %d", status);
```

Comment:
nit: have a separate method to delete tombstones for readability

Reply:
done